docker/build.sh

```diff
@@ -9,24 +9,132 @@
 # Jenkins build job should run with all options, for example,
 #   ./docker/build.sh jar -d push
 
-set -ex
-PROJECT=leonardo
+HELP_TEXT="$(cat <<EOF
+
+ Build the Leonardo code and docker images.
+   jar : build Leonardo jar
+   -d | --docker : (default: no action) provide either "build" or "push" to
+           build or push a docker image.  "push" will also perform build.
+   -r | --registry: (default: dockerhub) can be either "dockerhub" or "gcr".
+           Users of gcr should have the gcloud tool installed and configured.
+   -p | --project: set the project used at either dockerhub or with gcr
+           container registries.
+   -h | --help: print help text.
+
+ Examples:
+   Jenkins build job should run with all options, for example,
+     ./docker/build.sh jar -d push
+   To build the jar, the image, and push it to a gcr repository.
+     ./docker/build.sh jar -d build -r gcr --project "my-awesome-project"
+\t
+EOF
+)"
+
+# Enable strict evaluation semantics.
+set -e
+
+# Set default variables used while parsing command line options.
+TARGET="${TARGET:-leonardo}"
+GIT_BRANCH="${GIT_BRANCH:-$(git rev-parse --abbrev-ref HEAD)}"
+DOCKER_REGISTRY="dockerhub"  # Must be either "dockerhub" or "gcr"
+DOCKER_CMD=""
+ENV=${ENV:-""}  # if env is not set, push an image with branch name
+
+MAKE_JAR=false
+RUN_DOCKER=false
+PRINT_HELP=false
+
+
+if [ -z "$1" ]; then
+    echo "No argument supplied!"
+    echo "run '${0} -h' to see available arguments."
+    exit 1
+fi
+
+while [ "$1" != "" ]; do
+    case $1 in
+        jar)
+            MAKE_JAR=true
+            ;;
+        -d | --docker)
+            shift
+            echo "docker command = $1"
+            RUN_DOCKER=true
+            DOCKER_CMD="$1"
+            ;;
+        -r | --registry)
+            shift
+            echo "registry == $1"
+            DOCKER_REGISTRY=$1
+            ;;
+        -p | --project)
+            shift
+            echo "project = $1"
+            PROJECT=$1
+            ;;
+        -h | --help)
+            PRINT_HELP=true
+            ;;
+        *)
+            echo "Unrecognized argument '${1}'."
+            echo "run '${0} -h' to see available arguments."
+            if grep -Fq "=" <<< "${1}"; then
+                echo "note: separate args from flags with a space, not '='."
+            fi
+            exit 1
+            ;;
+    esac
+    shift
+done
+
+# Print help after all flags are parsed successfully
+if $PRINT_HELP; then
+  echo -e "${HELP_TEXT}"
+  exit 0
+fi
+
+# Configure script using arguments.
+if [[ $DOCKER_REGISTRY == "dockerhub" ]]; then
+  PROJECT="${PROJECT:-broadinstitute}"
```

Comment:
Heh, our Jenkins also sets `PROJECT=leonardo`. So `IMAGE` comes out as `leonardo/leonardo`. :)

Could we maybe call it `DOCKER_PROJECT` or something instead?